control_protocol.md

```diff
@@ -330,17 +330,490 @@ A Component MUST also offer a list of all possibly callable methods in accordanc
 For such a RPC message, the first content frame MUST consist in a JSON-RPC compatible content, for example a single request object or a batch of request objects.
 
 
+### List of methods
+
+List of methods defined by LECO.
+You MAY implement the optional methods of this list and you MUST implement the methods obligatory for your type of Component.
+All methods implemented in a Component MUST adhere to this list or MUST have a name different to any method in this list.
+
+
+#### Component
+
+Any Component, i.e. any participant in the LECO protocol, MUST offer the following methods.
+
+::: json
+{
+    "openrpc": "1.2.6",
+    "info": {
+        "title": "Component",
+        "version": "0.1.0"
+    },
+    "methods": [
+        {
+            "name": "pong",
+            "description": "Respond to a ping.",
+            "params": [],
+            "result": {
+                "name": "result",
+                "schema": {
+                    "type": "null"
+                },
+                "required": true
+            }
+        }
+    ],
+    "components": {
+        "schemas": {}
+    }
+}
+:::
+
+Any Component MAY offer ANY of the following methods.
+Components SHOULD offer ``shut_down``.
```

Comment:
I'm curious about the optional methods. What happens if some component calls another one's `shutdown`, but that is not implemented? Do we need some common fallback method? Do you get a not implemented error back or somesuch?

Reply:
> Do you get a not implemented error back or somesuch?

jsonrpc has a "method not found" error defined.

> I'm curious about the optional methods.

My reasoning is: There are some methods necessary, such that a Component fulfills it role (as a component, it has to be pingable), these are necessary.

Some features are nice to have, for example logging is not essential (maybe the Component does not have logging capabilities).

Reply:
Yes, I realise why optional methods can be useful. We should have a clear approach how do deal with that optionality, though -- there are several possibilities (do we error? ignore? publish/supply a list of implemented methods? how should the initiating party deal with that? ...)

Reply:
Components MUST adhere to openRPC, therefore they MUST offer a "rpc.discover" methods, which tells about all the implemented methods (the output is exactly that, what is part of this PR).

- [ ] Should we include this method in the Components list (together with "pong")?

If you try to call a method, a Component does not have, you get an appropriate error.

So, you (as a user) can either query beforehand and use only methods present, or you can try a method name and handle the error.

Reply:
> Should we include this method in the Components list (together with "pong")?

IMO, yes.

> So, you (as a user) can either query beforehand and use only methods present, or you can try a method name and handle the error.

It's more a LECO _implementor_ that has to deal with that. As we are designing for different implementations to talk to each other, I'm wondering if we should standardise the expectations around what happens with optional methods. The answer might as well be "no", or "use what `rpc.discover` gives you as you see fit".

Reply:
> "no", or "use what rpc.discover gives you as you see fit".

That's my stance 😁